compiler/semcall.nim

```diff
@@ -562,8 +562,62 @@ proc getCallLineInfo(n: PNode): TLineInfo =
     discard
   result = n.info
 
+proc inheritBindings(c: PContext, x: TCandidate, expectedType: PType): TIdTable =
+  ## Helper proc to inherit bound generic parameters from expectedType into a new TIdTable.
+  ## Returns existing bindings if 'inferGenericTypes' isn't in c.features
+  result = x.bindings
+  if inferGenericTypes notin c.features: return
+  if expectedType == nil or x.callee[0] == nil: return # required for inference
+
+  var
+    flatUnbound: seq[PType]
+    flatBound: seq[PType]
+  # seq[(result type, expected type)]
+  var typeStack = newSeq[(PType, PType)]()
```

Comment:
What are the effects of this computation on the overall compile-times?

Reply:
Not sure, asked the same before without receiving a response. Will copy and paste a million lines or so to test

Reply:
Using this code to generate two files, they both take about the same amount of time, but the inferred version takes an extra 30mb peakmem (cache cleared before each compile of course). Without the blocks I think it spends most of its time dealing with the symbol table.
<details>
<summary>Code</summary>

```nim


const copies = 10_000

block:
  var code = """{.experimental: "inferGenericTypes".}

type
  TestType[A, B, C, D, E, F, G, H, I, J] = object

proc giveData[A, B, C, D, E, F, G, H, I, J](): TestType[A, B, C, D, seq[E], F, G, H, I, seq[seq[J]]] = discard
"""

  for i in 0 ..< copies:
    code.add("""block:
  let x: TestType[int, float, int, char, seq[byte], int16, uint32, seq[float], int, seq[seq[int]]] = giveData()
""")


  writeFile("out1.nim", code)

block:
  var code = """type
  TestType[A, B, C, D, E, F, G, H, I, J] = object

proc giveData[A, B, C, D, E, F, G, H, I, J](): TestType[A, B, C, D, seq[E], F, G, H, I, seq[seq[J]]] = discard
"""

  for i in 0 ..< copies:
    code.add("""block:
  let x = giveData[int, float, int, char, seq[byte], int16, uint32, seq[float], int, seq[seq[int]]]()
""")


  writeFile("out2.nim", code)

```

</details>